### PR TITLE
API: Reformat output of groupby.describe (#4792)

### LINKED ISSUE #4792: describe on a groupby
I think expected result of this is with multiindex column...

```
df = pd.DataFrame({'PRICE': {pd.Timestamp('2011-01-06 10:59:05', tz=None): 24990,
  pd.Timestamp('2011-01-06 12:43:33', tz=None): 25499,
  pd.Timestamp('2011-01-06 12:54:09', tz=None): 25499},
 'VOLUME': {pd.Timestamp('2011-01-06 10:59:05', tz=None): 1500000000,
  pd.Timestamp('2011-01-06 12:43:33', tz=None): 5000000000,
  pd.Timestamp('2011-01-06 12:54:09', tz=None): 100000000}})


In [2]: g = df.groupby('PRICE')

In [3]: df.groupby('PRICE').describe()  # expected .unstack(1)
Out[3]: 
             PRICE        VOLUME
PRICE                           
24990 count      1  1.000000e+00
      mean   24990  1.500000e+09
      std      NaN           NaN
      min    24990  1.500000e+09
      25%    24990  1.500000e+09
      50%    24990  1.500000e+09
      75%    24990  1.500000e+09
      max    24990  1.500000e+09
25499 count      2  2.000000e+00
      mean   25499  2.550000e+09
      std        0  3.464823e+09
      min    25499  1.000000e+08
      25%    25499  1.325000e+09
      50%    25499  2.550000e+09
      75%    25499  3.775000e+09
      max    25499  5.000000e+09

```

cc #4740 @jreback 

I guess it just applies method/attribute from df:

```
In [4]: pd.DataFrame.foo = 1

In [5]: g.foo
Out[5]: 
PRICE
24990    1
25499    1
dtype: int64

In [6]: pd.DataFrame.bar = lambda x: 1

In [7]: g.bar()
Out[7]: 
PRICE
24990    1
25499    1
dtype: int64
```

should we just special case describe? :s

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #4792
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``

Doesn't look like this was address in a PR in 0.20, but the original issue works on master.

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -1504,6 +1504,31 @@ def test_frame_describe_tupleindex(self):
...
+                  pd.Timestamp('2011-01-06 12:43:33', tz=None): 25499,
+                  pd.Timestamp('2011-01-06 12:54:09', tz=None): 25499}
+        volumes = {pd.Timestamp('2011-01-06 10:59:05', tz=None): 1500000000,
+                   pd.Timestamp('2011-01-06 12:43:33', tz=None): 5000000000,
+                   pd.Timestamp('2011-01-06 12:54:09', tz=None): 100000000}
+        df = pd.DataFrame({'PRICE': prices,
+                           'VOLUME': volumes})
+        result = df.groupby('PRICE').describe()
+        expected_index = pd.MultiIndex(levels=[[24990, 25499],
+                                               ['count', 'mean', 'std',
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: instead of constructing this way

use concat with keys on the subframes

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
Actually I think the bug is still present. The issue that you want this

```
In [10]: df.groupby('PRICE').VOLUME.describe().unstack(1)
Out[10]: 
       count          mean           std           min           25%           50%           75%           max
PRICE                                                                                                         
24990    1.0  1.500000e+09           NaN  1.500000e+09  1.500000e+09  1.500000e+09  1.500000e+09  1.500000e+09
25499    2.0  2.550000e+09  3.464823e+09  1.000000e+08  1.325000e+09  2.550000e+09  3.775000e+09  5.000000e+09
```

rather than this
```
In [9]: df.groupby('PRICE').VOLUME.describe()
Out[9]: 
PRICE       
24990  count    1.000000e+00
       mean     1.500000e+09
       std               NaN
       min      1.500000e+09
       25%      1.500000e+09
       50%      1.500000e+09
       75%      1.500000e+09
       max      1.500000e+09
25499  count    2.000000e+00
       mean     2.550000e+09
       std      3.464823e+09
       min      1.000000e+08
       25%      1.325000e+09
       50%      2.550000e+09
       75%      3.775000e+09
       max      5.000000e+09
Name: VOLUME, dtype: float64
```

we do a similar unstack already with ``.ohlc``. 
```
In [8]: df.groupby('PRICE').VOLUME.ohlc()
Out[8]: 
             open        high         low       close
PRICE                                                
24990  1500000000  1500000000  1500000000  1500000000
25499  5000000000  5000000000   100000000   100000000
```

So each group gets a single row, while multi-columns are present for multiple aggregations.
multi-index are present for *multiple* groupers.

**Comment by mroeschke** [comment]
Ah that makes sense. Thanks for clarifying the expected output!

I've been poking into the code, and since each group goes through `apply()` and `describe()` returns the metrics labeled on the index, it tries to vertically concat the `describe()` results for each group:

```
(Pdb) values
[             VOLUME
count  1.000000e+00
mean   1.500000e+09
std             NaN
min    1.500000e+09
25%    1.500000e+09
50%    1.500000e+09
75%    1.500000e+09
max    1.500000e+09,              
              VOLUME
count  2.000000e+00
mean   2.550000e+09
std    3.464823e+09
min    1.000000e+08
25%    1.325000e+09
50%    2.550000e+09
75%    3.775000e+09
max    5.000000e+09]
```

I could add some logic saying if the indexes for each group are the same, concat on the index and transpose to the columns, but I think that'd be a pretty big change since it will probably affect all `groupby.apply()` functions. Or should we make a special case for `describe()`? Thoughts @jreback?

**Comment by jreback** [comment]
so you *do* just need to define ``.describe`` as a method on ``DataFrameGroupby`` and ``SeriesGroupby``, then you can do the ``.apply`` (followed by the unstack).

Right now we automagically just do any whitelisted function (including .describe) and they go thru the standard reshaping things. If you can figure out a generic way to do this great, but otherwise defining a method is fine.

**Comment by mroeschke** [comment]
Cool, thanks for the insight. 

I defined a new method for `groupby.describe()` and noticed in the process if `groupby(...,axis=1).describe()` is called, the transposing returns the desired results instead of unstacking. I also had to subsequently change a lot of the existing tests since the output has changed. Most notably the test for #14848 changed a lot.

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -366,6 +366,7 @@ Other API Changes
 - ``inplace`` arguments now require a boolean value, else a ``ValueError`` is thrown (:issue:`14189`)
 - ``pandas.api.types.is_datetime64_ns_dtype`` will now report ``True`` on a tz-aware dtype, similar to ``pandas.api.types.is_datetime64_any_dtype``
  - ``DataFrame.asof()`` will return a null filled ``Series`` instead the scalar ``NaN`` if a match is not found (:issue:`15118`)
+ - ``groupby.describe()`` now labels the `describe()` metrics in the column instead of the index (:issue:`4792`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: I think move this to a sub-section and show previous and current behavior.

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -1140,6 +1139,17 @@ def ohlc(self):
 
     @Substitution(name='groupby')
     @Appender(_doc_template)
+    def describe(self, **kwargs):
+        """
+        Provide summary statistics for each group, excluding NaN values
```
Comment (jreback) [inline pandas/core/groupby.py]: pls add a full doc-string, with named arguments. You might be able to simply add Series.describe and DataFrame.describe in the Notes section.

Comment (jreback) [inline pandas/core/groupby.py]: or as you do below just use the DataFrame doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
Added previous/current behavior in whatsnew, documentation to describe with `DataFrame.describe.__doc__`, and fixed other failing tests.

File: doc/source/whatsnew/v0.20.0.txt [hunk 4]
```diff
@@ -352,6 +352,68 @@ New Behavior:
    In [11]: index.memory_usage(deep=True)
    Out[11]: 260
 
+.. _whatsnew_0200.api_breaking.groupby_describe:
+
+Groupby Describe Formatting
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The output formatting of ``groupby.describe()`` now labels the ``describe()`` metrics in the columns instead of the index.
+This format is consistent with ``groupby.ohlc()`` (:issue:`4792`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: more to the point its consistent with how ``.agg()`` works.

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: most people prob don't know about ``.ohlc()`` :>

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -352,6 +352,68 @@ New Behavior:
...
+   In [3]: df.groupby('A').B.ohlc()
+   Out[3]: 
+   open  high  low  close
+   A                        
+   1     1     2    1      2
+   2     3     4    3      4
+
+New Behavior:
+
+.. code-block:: ipython
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: use an ipython-block here (so the code executes)

File: pandas/tests/groupby/test_groupby.py [hunk 6]
```diff
@@ -2994,16 +3021,25 @@ def test_non_cython_api(self):
         assert_frame_equal(result, expected)
 
         # describe
-        expected = DataFrame(dict(B=concat(
-            [df.loc[[0, 1], 'B'].describe(), df.loc[[2], 'B'].describe()],
-            keys=[1, 3])))
-        expected.index.names = ['A', None]
+        expected = pd.concat([(df[df.A == 1].B
+                                            .describe()
+                                            .to_frame()
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: This seems like lots of reshapings (this is current master)

```
In [6]: df
Out[6]: 
   A    B    C
0  1  2.0  foo
1  1  NaN  bar
2  3  NaN  baz

In [7]: df.groupby('A').describe()
Out[7]: 
           B
A           
1 count  1.0
  mean   2.0
  std    NaN
  min    2.0
  25%    2.0
  50%    2.0
  75%    2.0
  max    2.0
3 count  0.0
  mean   NaN
  std    NaN
  min    NaN
  25%    NaN
  50%    NaN
  75%    NaN
  max    NaN

In [8]: df.groupby('A').describe().unstack()
Out[8]: 
      B                                  
  count mean std  min  25%  50%  75%  max
A                                        
1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_groupby.py (hunk 6 above)
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby.py]: In this test, the `result = df.groupby('A').describe().unstack()` after `unstack()` was added to `groupby.describe()`. Shouldn't `expected` follow an independent path to the `result`?

Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: yes, but you can simply directly construct this result (as its 'simple' enough), just ``pd.DataFrame(.....)``

Comment (mroeschke) [inline pandas/tests/groupby/test_groupby.py]: Ah I see thanks! Agreed that my it my first edit probably included too much reshaping.

**Comment by mroeschke** [comment]
Replaced API example using `groupby.agg()` instead of `groupby.ohlc()`, fixed the code block, and simplified `expected` in `test_non_cython_api`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
hmm, something is wrong here. Its including the grouper.
```
In [1]: df = pd.DataFrame({'A': [1, 1, 2, 2], 'B': [1, 2, 3, 4]})

In [2]: df.groupby('A').describe()
Out[2]: 
      A                                        B                                          
  count mean  std  min  25%  50%  75%  max count mean       std  min   25%  50%   75%  max
A                                                                                         
1   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   2.0  1.5  0.707107  1.0  1.25  1.5  1.75  2.0
2   2.0  2.0  0.0  2.0  2.0  2.0  2.0  2.0   2.0  3.5  0.707107  3.0  3.25  3.5  3.75  4.0

In [3]: df.groupby('A').agg(['mean', 'std'])
Out[3]: 
     B          
  mean       std
A               
1  1.5  0.707107
2  3.5  0.707107
```

**Comment by mroeschke** [comment]
It looks like the grouper is included when using `groupby.apply()`:

```
In [4]: pd.__version__
Out[4]: u'0.19.2' #not current master

In [5]: df.groupby('A').apply(lambda x: x.describe()) 
Out[5]:
           A         B
A
1 count  2.0  2.000000
  mean   1.0  1.500000
  std    0.0  0.707107
  min    1.0  1.000000
  25%    1.0  1.250000
  50%    1.0  1.500000
  75%    1.0  1.750000
  max    1.0  2.000000
2 count  2.0  2.000000
  mean   2.0  3.500000
  std    0.0  0.707107
  min    2.0  3.000000
  25%    2.0  3.250000
  50%    2.0  3.500000
  75%    2.0  3.750000
  max    2.0  4.000000

```

Can be seen with other functions as well:

```
In [11]: df.groupby('A').apply(np.mean) #not idiomatic but should be similar to  df.groupby('A').mean()
Out[11]:
     A    B
A
1  1.0  1.5
2  2.0  3.5
```

Is this a known issue?

**Comment by jreback** [comment]
@mroeschke so by-definition this is what apply does.

you can use ``._set_group_selection()`` to avoid this problem.

**Comment by mroeschke** [comment]
Ah thanks for clarifying that @jreback. Will edit to use `._set_group_selection()` tonight.

**Comment by mroeschke** [comment]
Added `_set_group_selection()` to prevent the grouper from being included and resolved conflict with whatsnew

**Comment by jreback** [comment]
thanks!

keep em coming!

**Pull request closed without merging**